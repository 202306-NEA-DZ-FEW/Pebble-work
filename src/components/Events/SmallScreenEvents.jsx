import React, { useState, useEffect, useRef } from "react";
import Calendar from "@/components/Filter/Calendar";
import MobileCard from "./MobileCard";

import styles from "@/styles/Events.module.css";

import FilterByType from "@/components/Filter/FilterByType";
import FirestoreLocation from "@/components/Filter/FirestoreLocation";

import { usePagination } from "@/components/Pagination/Pagination";

const SmallScreenEvents = (user) => {
    // State variables

    const [selectedTypes, setSelectedTypes] = useState([]);
    const [inputValue1, setInputValue1] = useState("");
    const [isCalendarOpen, setCalendarOpen] = useState(false);

    const [filteredTypes, setFilteredTypes] = useState([]);
    const [events, setEvents] = useState([]);
    const [filteredEvents, setFilteredEvents] = useState([]);
    const [CalendarEvents, setCalendarEvents] = useState([]);
    const [selectedDate, setSelectedDate] = useState(null);
    const [resetLocation, setResetLocation] = useState(false);
    const [resetDays, setResetDays] = useState(false);
    const dropdownRef = useRef(null);
    const { currentPage, totalPages, currentItems, setCurrentPage } =
        usePagination(1, 5, filteredEvents);

    const handleLocationInputChange = (value) => {
        setInputValue1(value);
    };

    const handleClickOutside = (event) => {
        if (
            dropdownRef.current &&
            !dropdownRef.current.contains(event.target)
        ) {
            setFilteredTypes(false);
        }
    };
    useEffect(() => {
        window.addEventListener("click", handleClickOutside);
        return () => {
            window.removeEventListener("click", handleClickOutside);
        };
    }, []);

    // Handle test click
    const handleCalendarClick = () => {
        setCalendarOpen(!isCalendarOpen);
    };

    // Fetch events from Firebase
    useEffect(() => {
        const fetchEvents = async () => {
            const types = selectedTypes ? selectedTypes.join(",") : "";
            const dates = selectedDate ? selectedDate.join(",") : "";
            const response = await fetch(
                `/api/events?type=${types}&location=${inputValue1}&date=${dates}`
            );
            const eventsData = await response.json();

            setEvents(eventsData);
            setCalendarEvents(eventsData);
        };

        fetchEvents();
    }, [selectedTypes, inputValue1, selectedDate]);

    useEffect(() => {
        const applyFilters = () => {
            let filteredEvents = events; // Use all events, not just current page

            // Apply type filter
            if (filteredTypes.length > 0) {
                filteredEvents = filteredEvents.filter((event) =>
                    filteredTypes.includes(event.type)
                );
            }

            // Apply location filter
            if (inputValue1) {
                filteredEvents = filteredEvents.filter(
                    (event) => event.location === inputValue1
                );
            }

            // Apply date filter
            if (selectedDate && selectedDate.length > 0) {
                filteredEvents = filteredEvents?.filter((event) =>
                    selectedDate.includes(event.date)
                );
            }

            setFilteredEvents(filteredEvents);
        };

        applyFilters();
    }, [events, selectedDate, inputValue1, filteredTypes]); // Depend on all filters

    const resetEvents = () => {
        setSelectedTypes([]);
        setInputValue1("");
        setSelectedDate(null);
        setFilteredTypes([]);
        setFilteredEvents([]);
        setCalendarEvents([]);
        setResetLocation(true);
        setResetDays([]);
    };
    const handleInputDelete = () => {
        setInputValue1("");
    };
    const handleDateChange = (dates) => {
        setSelectedDate(
            dates.map((date) => new Date(date).toISOString().split("T")[0])
        );
    };
    return (
        <>
            <main
                className={` flex flex-col justify-center items-center overflow-scroll`}
            >
                <div>
                    <h1>Welcome, {user.name}!</h1>
                    <p>This is the events page</p>
                </div>
                <button
                    onClick={resetEvents}
                    className={` w-[52px] bg-blue-400 text-white text-[10px] hover:bg-blue-500 rounded-[4px] h-[16px] ml-auto mr-2`}
                >
                    All events
                </button>
                <div className={`flex flex-col-reverse`}>
                    <div className={`h-[750px]`}>
                        <ul
                            className={` flex flex-row pb-5 justify-center flex-wrap items gap-4`}
                        >
                            {currentItems.map((event) => {
                                return (
                                    <MobileCard
                                        eventId={event.id}
                                        key={event.id}
                                        title={event.title}
                                        type={event.type}
                                        image={
                                            event?.image || "/event_image.png"
                                        }
                                        location={event.location}
                                        description={event.description}
                                        organizer={event.organizer}
                                        time={event.time}
                                        date={event.date}
                                    />
                                );
                            })}
                            {(inputValue1 || filteredTypes.length > 0) &&
                                filteredEvents.length === 0 && (
                                    <p className='text-red-500 text-center'>
                                        No events found for this date and
                                        location
                                    </p>
                                )}

                            {selectedDate && CalendarEvents.length === 0 && (
                                <p className='text-red-500 text-center'>
                                    No events found for this date
                                </p>
                            )}
                        </ul>
                        <div className='flex justify-center pb-2 gap-8 text-black'>
                            {Array.from(
                                { length: totalPages },
                                (_, i) => i + 1
                            ).map((pageNumber) => (
                                <button
                                    key={pageNumber}
                                    onClick={() => setCurrentPage(pageNumber)}
                                    disabled={pageNumber === currentPage}
                                >
                                    {pageNumber}
                                </button>
                            ))}
                        </div>
                    </div>
                    <div className='flex bg-white z-10 flex-row items-center justify-evenly text-black'>
                        <div>
                            <button
                                className='sm:hidden'
                                onClick={handleCalendarClick}
                            >
                                Dates
                            </button>
                            {isCalendarOpen && (
                                <div
                                    style={{
                                        animation: `${
                                            isCalendarOpen
                                                ? `${styles.fadeIn} 0.7s ease-in-out`
                                                : ""
                                        }`,
                                    }}
                                    className={`text-center ${
                                        isCalendarOpen ? "open" : ""
                                    } ${
                                        styles.calendarContainer
                                    } shadow-inner rounded-[4px] z-10 bg-white`}
                                >
                                    <button
                                        className='btn rounded-[8px]'
                                        onClick={() => setCalendarOpen(false)}
                                    >
                                        Close
                                    </button>
                                    <Calendar
                                        resetDays={resetDays}
                                        checkEvents={handleDateChange}
                                    />
                                </div>
                            )}
                        </div>
                        <div className='h-66'>
                            <FirestoreLocation
                                onInputChange={handleLocationInputChange}
                                resetLocation={resetLocation}
                                setResetLocation={setResetLocation}
                                onInputDelete={handleInputDelete}
                            />
                        </div>
                        <FilterByType
                            ref={dropdownRef}
                            setFilteredTypes={setFilteredTypes}
                            selectedTypes={selectedTypes}
                            setSelectedTypes={setSelectedTypes}
                        />
                    </div>
                </div>
            </main>
        </>
    );
};

export default SmallScreenEvents;
