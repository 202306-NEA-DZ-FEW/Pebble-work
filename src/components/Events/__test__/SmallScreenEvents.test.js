import React from "react";

import SmallScreenEvents from "../SmallScreenEvents";
import { render, cleanup } from "@testing-library/react";
afterEach(cleanup);
describe("SmallScreenEvents", () => {
    it("renders correctly", () => {
        const user = { name: "Test User" };

        const { asFragment } = render(<SmallScreenEvents user={user} />);

        expect(asFragment()).toMatchSnapshot();
    });
});
