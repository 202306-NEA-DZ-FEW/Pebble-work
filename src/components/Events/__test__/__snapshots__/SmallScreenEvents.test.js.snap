// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmallScreenEvents component renders correctly 1`] = `
<main
  className=" flex flex-col justify-center items-center overflow-scroll"
>
  <div>
    <h1>
      Welcome, 
      !
    </h1>
    <p>
      This is the events page
    </p>
  </div>
  <button
    className=" w-[52px] bg-blue-400 text-white text-[10px] hover:bg-blue-500 rounded-[4px] h-[16px] ml-auto mr-2"
    onClick={[Function]}
  >
    All events
  </button>
  <div
    className="flex flex-col-reverse"
  >
    <div
      className="h-[750px]"
    >
      <ul
        className=" flex flex-row pb-5 justify-center flex-wrap items gap-4"
      />
      <div
        className="flex justify-center pb-2 gap-8 text-black"
      />
    </div>
    <div
      className="flex bg-white z-10 flex-row items-center justify-evenly text-black"
    >
      <div>
        <button
          className="sm:hidden"
          onClick={[Function]}
        >
          Dates
        </button>
      </div>
      <div
        className="h-66"
      >
        <button
          className="btn sm:bg-blue-500   hover:bg-gray-400 sm:text-center flex lg:items-center sm:gap-0 gap-2 sm:justify-center lg:w-[281px] lg:h-[52px] sm:w-[149px] sm:h-[30px] text-white font-[500] lg:text-[18px] lg:tracking-[0.10px] sm:text-[12px] sm:px-4 sm:py-2 rounded-lg"
          onClick={[Function]}
        >
          Location
        </button>
        <dialog
          className="modal rounded-tr rounded-br"
          id="my_modal_1"
        >
          <div
            className="modal-box"
          >
            <div
              className="flex justify-center"
            >
              <input
                className="mt-2 p-2 border rounded"
                onChange={[Function]}
                placeholder="Search city..."
                type="text"
                value=""
              />
            </div>
            <h3
              className="font-bold text-lg mb-0"
            >
              cities 
            </h3>
            <div
              className=" mt-2  "
            >
              <ul
                className="list-none information h-[200px]"
              />
              <div
                className="modal-action absolute top-7 right-3 mt-0"
              >
                <form
                  method="dialog"
                >
                  <button
                    className=""
                    onClick={[Function]}
                  >
                    <svg
                      fill="currentColor"
                      fillRule="evenodd"
                      height={20}
                      stroke="currentColor"
                      strokeWidth="0"
                      style={
                        {
                          "color": undefined,
                        }
                      }
                      viewBox="0 0 1024 1024"
                      width={20}
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M799.855 166.312c.023.007.043.018.084.059l57.69 57.69c.041.041.052.06.059.084a.118.118 0 0 1 0 .069c-.007.023-.018.042-.059.083L569.926 512l287.703 287.703c.041.04.052.06.059.083a.118.118 0 0 1 0 .07c-.007.022-.018.042-.059.083l-57.69 57.69c-.041.041-.06.052-.084.059a.118.118 0 0 1-.069 0c-.023-.007-.042-.018-.083-.059L512 569.926 224.297 857.629c-.04.041-.06.052-.083.059a.118.118 0 0 1-.07 0c-.022-.007-.042-.018-.083-.059l-57.69-57.69c-.041-.041-.052-.06-.059-.084a.118.118 0 0 1 0-.069c.007-.023.018-.042.059-.083L454.073 512 166.371 224.297c-.041-.04-.052-.06-.059-.083a.118.118 0 0 1 0-.07c.007-.022.018-.042.059-.083l57.69-57.69c.041-.041.06-.052.084-.059a.118.118 0 0 1 .069 0c.023.007.042.018.083.059L512 454.073l287.703-287.702c.04-.041.06-.052.083-.059a.118.118 0 0 1 .07 0Z"
                      />
                    </svg>
                  </button>
                </form>
              </div>
            </div>
          </div>
        </dialog>
      </div>
      <div>
        <button
          className="sm:hidden"
          onClick={[Function]}
        >
          Interests
        </button>
      </div>
    </div>
  </div>
</main>
`;
